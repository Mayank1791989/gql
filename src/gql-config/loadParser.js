/* @flow */
/* @babel-flow-runtime-enable */
import {
  type PkgConfig,
  type SchemaParserLoadedPkg,
  type QueryParserLoadedPkg,
} from './types';
import PkgUtils from './PkgUtils';

import { reify, Type } from 'flow-runtime';

export function loadQueryParser(
  parser: ?PkgConfig,
  extraParserOptions: $FixMe,
  pkgUtils: PkgUtils,
): QueryParserLoadedPkg {
  return pkgUtils.load({
    prefix: 'gql-query-parser',
    pkgConfig: parser || 'default',
    extraPkgOptions: extraParserOptions,
    corePkgs: ['gql-query-parser-default', 'gql-query-parser-embedded-queries'],
    validate(parserModule): QueryParserLoadedPkg {
      // const ParserType = (reify: Type<QueryParserLoadedPkg>);
      // ParserType.assert(parserModule);
      return (parserModule: $FixMe);
    },
  });
}

export function loadSchemaParser(
  parser: ?PkgConfig,
  extraParserOptions: $FixMe,
  pkgUtils: PkgUtils,
): SchemaParserLoadedPkg {
  return pkgUtils.load({
    prefix: 'gql-schema-parser',
    pkgConfig: parser || 'default',
    corePkgs: ['gql-schema-parser-default'],
    extraPkgOptions: extraParserOptions,
    validate(parserModule): SchemaParserLoadedPkg {
      // const ParserType = (reify: Type<SchemaParserLoadedPkg>);
      // ParserType.assert(parserModule);
      return (parserModule: $FixMe);
    },
  });
}

// export function loadResolverParser(
//   parser: ParserPkg,
//   configPath: string,
// ): [Class<IResolverParser>, Options] {
//   return loadParser({
//     parser,
//     prefix: 'gql-resolver-parser',
//     coreParsers: [],
//     configPath,
//     validate(parserModule): Class<IResolverParser> {
//       const ParserType = (reify: Type<IResolverParser>);
//       ParserType.assert(parserModule);
//       return (parserModule: $FixMe);
//     },
//   });
// }

// function loadParser<
//   TParser: IQueryParser | ISchemaParser | IResolverParser,
// >(params: {
//   parser: ParserPkg,
//   prefix: string,
//   configPath: string,
//   coreParsers: Array<string>,
//   validate: (parserModule: mixed) => Class<TParser>,
// }): [Class<TParser>, Options] {
//   const [parserPkg, options] = normalizePkgConfig(params.prefix, params.parser);

//   const [pkg, dir] = params.coreParsers.includes(parserPkg)
//     ? [`./${parserPkg}`, GQL_MODULE_DIR]
//     : [parserPkg, params.configPath];

//   try {
//     const parserModule = importModule(pkg, dir);
//     const parser = params.validate(parserModule);
//     return [parser, options];
//   } catch (err) {
//     if (err.code === 'MODULE_NOT_FOUND') {
//       throw new Error(
//         `PARSER_PKG_NOT_FOUND: Parser '${pkg}' not found relative to '${dir}'`,
//       );
//     }

//     if (err.name === 'RuntimeTypeError') {
//       throw new Error(
//         `PARSER_PKG_INVALID: There are errors in parser '${pkg}'\n\n${
//           err.message
//         }`,
//       );
//     }

//     throw err;
//   }
// }
