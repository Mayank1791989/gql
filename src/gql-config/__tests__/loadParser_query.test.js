/* @flow */
import { loadQueryParser } from '../loadParser';

describe('load core-parsers:', () => {
  [
    'gql-query-parser-default',
    'default',
    'gql-query-parser-embedded-queries',
    'embedded-queries',
  ].forEach(parser => {
    test(`using full package name =${parser}`, () => {
      expect(loadQueryParser(parser, '')).toBeDefined();
    });
  });

  ['default', 'embedded-queries'].forEach(parser => {
    test(`using short package name =${parser}`, () => {
      expect(loadQueryParser(parser, '')).toBeDefined();
    });
  });
});

describe('missing-parser-pkg', () => {
  [
    'gql-query-parser-unknown',
    './custom-parser',
    '@scope/custom-parser',
    'custom_node_module_parser',
  ].forEach(parserPkg => {
    expect(() => loadQueryParser(parserPkg, '')).toThrowErrorMatchingSnapshot();
  });
});
