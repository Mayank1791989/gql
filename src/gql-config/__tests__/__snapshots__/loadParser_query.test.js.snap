// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"PARSER_PKG_NOT_FOUND: Parser 'gql-query-parser-unknown' not found relative to ''"`;

exports[` 2`] = `"PARSER_PKG_NOT_FOUND: Parser './custom-parser' not found relative to ''"`;

exports[` 3`] = `"PARSER_PKG_NOT_FOUND: Parser '@scope/gql-query-parser-custom-parser' not found relative to ''"`;

exports[` 4`] = `"PARSER_PKG_NOT_FOUND: Parser 'gql-query-parser-custom_node_module_parser' not found relative to ''"`;
